### ISSUE TITLE
Race condition in github workflow 

### ISSUE THREAD

**Issue by hlinnaka** [neondatabase/neon#2647]
The PR workflow runs the "promote-images" job, which does this for each of the images:
```
Run MANIFEST=$(aws ecr batch-get-image --repository-name neon --image-ids imageTag=$GITHUB_RUN_ID --query 'images[].imageManifest' --output text) && aws ecr put-image --repository-name neon --image-tag latest --image-manifest "$MANIFEST"
```
And then, on commits to 'main' but not on PRs, the 'push-docker-hub' stage pulls the image back from ECR:
```
crane pull 369495373322.dkr.ecr.eu-central-1.amazonaws.com/neon:latest neon
```
There a race condition, if a PR workflow and the deployment run at the same time. Tthe PR workflow could replace the image with the tag 'latest', while the 'main' deployment is between the two steps above, and it will incorrectly deploy the image from the PR instead.

We saw this happen on staging, see slack conversation below:

### Conversation we had on slack (slightly edited)

Alexey:
I see a lot of errors like
> could not execute operation: could not start the compute node: db error: FATAL: could not read block 0 in rel 1664/0/1260.0 from page server at lsn 0/065B9FA8 DETAIL: page server returned error: WAL redo process closed its stderr unexpectedly Caused by: FATAL: could not read block 0 in rel 1664/0/1260.0 from page server at lsn 0/065B9FA8 DETAIL: page server returned error: WAL redo process closed its stderr unexpectedly

Sergey:
> Oct 18 16:26:11 zenith-us-stage-ps-2.local pageserver[1734545]: 2022-10-18T16:26:11.462418Z ERROR handle_pagerequests{tenant_id=abfcf56125937829876a641640afedd8 timeline_id=a398ca9109804294f414ea72e6589526}:handle_get_page_at_lsn_request{rel=1664/0/1260 blkno=0 req_lsn=0/2597488}: wal-redo-postgres: 2022-10-18 16:26:11.462 UTC [1736678] FATAL:  could not load library "/usr/local/v14/lib/postgresql/neon_walredo.so": /usr/local/v14/lib/postgresql/neon_walredo.so: undefined symbol: seccomp_load_rules

Correlates by time with https://github.com/neondatabase/neon/pull/2633

This errors at 16:07 UTC
If this would be production I would rollback #2633, but this is staging

Konstantin:
```
admin@zenith-us-stage-ps-2:~$ ls -l /usr/local/v14/bin/postgres
-rwxr-xr-x 1 root root 90752928 Oct 18 15:55 /usr/local/v14/bin/postgres
```
Sergey:
```
admin@zenith-us-stage-ps-2:~$ ls -l /usr/local/v14/lib/postgresql/neon_walredo.so 
-rwxr-xr-x 1 root root 992144 Oct 18 15:34 /usr/local/v14/lib/postgresql/neon_walredo.so
```
Konstantin:
```
admin@zenith-us-stage-ps-2:~$ nm -s /usr/local/v14/bin/postgres | fgrep seccomp_load_rules
0000000000371230 T seccomp_load_rules
000000000037145b t seccomp_load_rules.cold
```
I wonder how are we upgrading Postgres? Do we stop pageserver, then copy files and then start server?
May be for some reasons old wal-redo process is still running, pinning neon_walredo.so  library in memory.In any case, as far as I undesrand - now the problem is "solved": no more wal-redo errors?

Alexey:
> I wonder how are we upgrading Postgres? Do we stop pageserver, then copy files and then start server?

IIRC, we just unpack the archive with all binaries, then restart the service

Sergey:
Yes, looks like it fixed itself at 16:35 UTC with next deploy

Alexey:
> In any case, as far as I undesrand - now the problem is "solved": no more wal-redo errors?
I think so, the problem was only during deploy

Yet, I see two more new stuck projects from today:
* https://console.stage.neon.tech/admin/projects/red-hall-619366
* https://console.stage.neon.tech/admin/projects/raspy-firefly-341523

Both with a weird error:
```
Tenant 6d2ec7bcf0cdb2e4f0d3bfa53479bfa8 not found in the local state
```

Sergey:
> I think so, the problem was only during deploy
Incorrect, it was whole time between two deploys

Alexey:
Not sure this is related, but both had zero problems until today

Heikki:
That file, neon_walredo.so is created in PR https://github.com/neondatabase/neon/pull/2632. It's not present on 'main' branch.My hunch is that those errors are coming from the end-to-end tests of that PR. Do errors in end-to-end tests on PRs affect the SLO? And they also add noise to the staging logs, I guess.

Konstantin:
So we are unpacking old archive overring old binaries? Or we are unpacking it in some other directory nd then moving it to taget location? I am not sure that overriding existed boundaries is good idea.

Alexey:
> Do errors in end-to-end tests on PRs affect the SLO?
No, for PRs e2e is running in the CI containers
We only run e2e againts staging / prod during deploy of cloud commits from main and release

Sergey:
This is how we deploy, need to check what ansible exactly do https://github.com/neondatabase/neon/blob/8fbe437768e2545855a0f928794abc679b7f9d0b/.github/ansible/deploy.yaml#L23

Heikki:
Here's similar errors on that PR's e2e test:
https://github.com/neondatabase/cloud/actions/runs/3274962365/jobs/5389466000
Many of the e2e tests actually failed to run.It does look like some kind of mixup between that the end-to-end tests of that PR, and the deployment.
Hmm, that PR workflow run the "promote-images" job, which does this for each of the images:
```
Run MANIFEST=$(aws ecr batch-get-image --repository-name neon --image-ids imageTag=$GITHUB_RUN_ID --query 'images[].imageManifest' --output text) && aws ecr put-image --repository-name neon --image-tag latest --image-manifest "$MANIFEST"
```
And then, on commits to 'main' but not on PRs, the 'push-docker-hub' stage pulls the image back from ECR:
```
crane pull 369495373322.dkr.ecr.eu-central-1.amazonaws.com/neon:latest neon
```
Isn't there a race condition, if a PR workflow and the deployment run at the same time? If I understand this correctly, the PR workflow would replace the image with the tag 'latest'

I think the intention was that 'promote-images' would not run on PRs, only on pushes to 'main'

**Comment by zoete** [comment]
> 
> I think the intention was that 'promote-images' would not run on PRs, only on pushes to 'main'

Correct, it was left in the current situation pending more specific changes being implemented.

**Comment by lubennikovaav** [comment]
so the fix would be

```
diff --git a/.github/workflows/build_and_test.yml b/.github/workflows/build_and_test.yml
index d90455cc..29ee73b7 100644
--- a/.github/workflows/build_and_test.yml
+++ b/.github/workflows/build_and_test.yml
@@ -565,7 +565,9 @@ jobs:
   promote-images:
     runs-on: dev
     needs: [ neon-image, compute-node-image, compute-node-image-v14, compute-node-image-v15, compute-tools-image ]
-    if: github.event_name != 'workflow_dispatch'
+    if: |
+      (github.ref_name == 'main' || github.ref_name == 'release') &&
+      github.event_name != 'workflow_dispatch'
     container: amazon/aws-cli
     strategy:
       fail-fast: false
       ```
IIUC, it still has race between `main` and `release`. How can we fix it?
